docs/book/src/developer/tests.md

```diff
@@ -0,0 +1,28 @@
+# Running Cluster API tests
+
+## Unit tests
+
+Unit tests run very quickly. They don't require any additional services and can be run using default go tools or through
+the `test` make target, e.g. `make test`.
+
+## Integration tests
+
+Integration tests use a real cluster and real dependencies to run tests. The dependencies are managed manually and are
+not meant to be run locally. See `scripts/ci-integration.sh` for more details.
+
+## End-to-end tests
+
+The end-to-end tests are similar to the integration tests except they are designed to manage dependencies for you and 
+have a more complete test using a docker provider to test the Cluster API mechanisms. The end-to-end tests run best on 
+CI, but work locally if you're looking for early signal.
+
+To run the tests, first, build the images needed for Cluster API from your local checkout. Then build the docker 
+provider manager image. After that you can run the tests directly from make. Here are the commands to do the above steps. 
+
+```
+make docker-build REGISTRY=gcr.io/k8s-staging-cluster-api
+make -C test/infrastructure/docker docker-build REGISTRY=gcr.io/k8s-staging-capi-docker
+make -C test/infrastructure/docker run-e2e
```

Comment:
Would it be easier to default to the values used in CI and tell users to run bash script similar to what you are doing above for the integration tests?

Reply:
hmm, I think this would end up with the user modifying the script if they, say, didn't want to rebuild the images every time. I think the flexibility should lie with the user and not with CI.

Would this be more clear if I said: After the images are built they only need to be rebuilt to pick up changes for any given manager. For example, if the bootstrap manager changes then the bootstrap image must be rebuilt but not the other two.

Reply:
sgtm